File: agent-runtime/tests/system_e2e/knowledge_human_review.js
"use strict";
let token = location.hash.slice(1);
history.replaceState(null, "", "/");
let nonce = null, caseId = null, stopped = false, busy = false, polling = false, generation = 0;
const byId = id => document.getElementById(id);
const labels = {faithful: "忠于原文，没有误导", relevant: "与问题相关",
  sufficientForInitialAnswer: "足以形成初步回答，显式条件未遗漏", useful: "对理解问题有实际帮助"};
for (const [name, label] of Object.entries(labels)) {
  const field = document.createElement("fieldset"), legend = document.createElement("legend");
  legend.textContent = label; field.append(legend);
  for (const [value, text] of [["true", "是"], ["false", "否"]]) {
    const wrapper = document.createElement("label"), input = document.createElement("input");
    input.type = "radio"; input.name = name; input.value = value; input.required = true;
    wrapper.append(input, document.createTextNode(text)); field.append(wrapper);
  }
  byId("rubric").append(field);
}
function clearPacket() {
  for (const id of ["question", "answer", "points", "evidence", "case"]) byId(id).replaceChildren();
  byId("form").reset(); byId("review").hidden = true; nonce = caseId = null;
}
async function request(path, body) {
  const response = await fetch(path, {method: body === undefined ? "GET" : "POST", cache: "no-store",
    credentials: "omit", headers: {"X-Review-Token": token, ...(body === undefined ? {} : {"Content-Type": "application/json"})},
    ...(body === undefined ? {} : {body: JSON.stringify(body)})});
  if (!response.ok) throw new Error("评审请求未接受，请勿刷新或重复运行。");
  return response.json();
}
function textBlock(parent, text) {
  const node = document.createElement("pre"); node.textContent = text; parent.append(node);
}
async function poll() {
  if (stopped || busy || polling) return;
  polling = true; const currentGeneration = generation;
  try {
    const state = await request("/state");
    if (stopped || busy || generation !== currentGeneration) return;
    byId("ready").hidden = state.status !== "ready_required";
    if (state.status === "review" && state.nonce !== nonce) {
      clearPacket(); nonce = state.nonce; const item = state.packet; caseId = item.caseId;
      byId("case").textContent = caseId; byId("question").textContent = item.question;
      byId("answer").textContent = item.answerSummary;
      for (const point of item.points) textBlock(byId("points"), point.quote + "\n" + JSON.stringify(point.citation, null, 2));
      for (const evidence of item.evidence) {
        const details = document.createElement("details"), summary = document.createElement("summary");
        summary.textContent = [evidence.evidenceRef, evidence.title, evidence.documentNumber, evidence.writtenDate].filter(Boolean).join(" · ");
        details.append(summary); textBlock(details, evidence.content); byId("evidence").append(details);
      }
      byId("review").hidden = false;
    } else if (state.status !== "review") clearPacket();
    byId("status").textContent = ({ready_required: "尚未执行：请确认本人已准备好逐题评价。",
      waiting: "已就绪，等待本题真实回答。", review: "请阅读并评价当前题。任何不通过均会停止本批。", closed: "本批已停止。"})[state.status];
    if (state.status === "closed") {stopped = true; token = "";}
  } catch (error) {
    if (!stopped && !busy && generation === currentGeneration) {
      clearPacket(); stopped = true; token = ""; byId("status").textContent = error.message;
    }
  } finally {polling = false;}
}
byId("ready").onclick = async () => {
  busy = true; generation++; byId("ready").disabled = true;
  try {await request("/ready", {ready: true});}
  catch (error) {byId("status").textContent = error.message; stopped = true;}
  finally {busy = false; await poll();}
};
byId("form").onsubmit = async event => {
  event.preventDefault(); if (busy || !nonce) return;
  const data = new FormData(byId("form")), decision = {caseId, nonce, acknowledged: byId("ack").checked, reason: byId("reason").value};
  for (const name of Object.keys(labels)) decision[name] = data.get(name) === "true";
  if (Object.keys(labels).every(name => decision[name]) !== (decision.reason === "none")) {
    byId("status").textContent = "原因与四项评价不一致，请检查。"; return;
  }
  busy = true; generation++;
  try {await request("/decision", decision); clearPacket();}
  catch (error) {byId("status").textContent = error.message; stopped = true; clearPacket();}
  finally {busy = false; await poll();}
};
byId("close").onclick = async () => {
  busy = true; generation++;
  try {await request("/close", {close: true});} finally {
    stopped = true; token = ""; clearPacket(); byId("ready").hidden = true; byId("status").textContent = "本批已停止。";
  }
};
window.addEventListener("pagehide", () => {
  if (token && !stopped) fetch("/close", {method: "POST", keepalive: true, cache: "no-store", credentials: "omit",
    headers: {"X-Review-Token": token, "Content-Type": "application/json"}, body: '{"close":true}'}).catch(() => {});
  generation++; clearPacket(); token = ""; stopped = true;
});
if (!token) {stopped = true; byId("status").textContent = "缺少本次临时会话。请从受控执行器打开，不要手动刷新。";}
else {poll(); setInterval(poll, 1000);}

File: agent-runtime/tests/system_e2e/knowledge_human_review.py
"""Ephemeral, loopback-only human appraisal. Never records displayed content."""
from __future__ import annotations

import hashlib
import hmac
import json
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
import secrets
import threading
import time
from typing import Any, cast
import webbrowser

FIELDS = ("faithful", "relevant", "sufficientForInitialAnswer", "useful")
REASONS = {"none", "quote_context", "relevance", "coverage", "gold_issue"}
MAX_PACKET_BYTES = 131072
MAX_POST_BYTES = 2048


def canonical(value: Any) -> bytes:
    return json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":"), allow_nan=False).encode()


def sha(value: Any) -> str:
    return hashlib.sha256(canonical(value)).hexdigest()


def exact_json(raw: bytes) -> Any:
    def pairs(items: list[tuple[str, Any]]) -> dict[str, Any]:
        result: dict[str, Any] = {}
        for key, value in items:
            if key in result:
                raise ValueError("duplicate_key")
            result[key] = value
        return result

    def nonfinite(_: str) -> None:
        raise ValueError("nonfinite")

    return json.loads(raw, object_pairs_hook=pairs, parse_constant=nonfinite)


def packet(case: dict[str, Any], response: dict[str, Any], summary_input: Any) -> dict[str, Any]:
    """Project only the actual public answer and same-request policy-visible input."""
    result = response.get("result")
    if response.get("status") != "success" or not isinstance(result, dict) or summary_input is None:
        raise ValueError("review.answer_unavailable")
    if summary_input.question != case["question"]:
        raise ValueError("review.question_mismatch")
    points = result.get("points")
    if not isinstance(points, list) or not 1 <= len(points) <= 8 or not 1 <= len(summary_input.evidence) <= 8:
        raise ValueError("review.shape_invalid")
    projected = []
    for point in points:
        citation = point.get("citation") if isinstance(point, dict) else None
        if not isinstance(citation, dict) or not isinstance(point.get("quote"), str):
            raise ValueError("review.point_invalid")
        projected.append(dict(quote=point["quote"], citation={k: citation[k] for k in
            ("evidenceId", "domainIds", "title", "documentNumber", "writtenDate") if k in citation}))
    value = dict(caseId=case["caseId"], question=case["question"], answerSummary=result.get("answerSummary"),
        points=projected, evidence=[dict(evidenceRef=e.evidence_ref, content=e.content,
            domainIds=e.domain_ids, title=e.title, documentNumber=e.document_number, writtenDate=e.written_date)
            for e in summary_input.evidence])
    if not isinstance(value["answerSummary"], str) or len(canonical(value)) > MAX_PACKET_BYTES:
        raise ValueError("review.packet_invalid")
    return value


def verdict(value: Any, case_id: str, nonce: str) -> dict[str, Any]:
    required = {"caseId", "nonce", "acknowledged", "reason", *FIELDS}
    if not isinstance(value, dict) or set(value) != required:
        raise ValueError("review.verdict_shape")
    if value["caseId"] != case_id or value["nonce"] != nonce or value["acknowledged"] is not True:
        raise ValueError("review.verdict_binding")
    if (any(type(value[k]) is not bool for k in FIELDS)
            or not isinstance(value["reason"], str) or value["reason"] not in REASONS):
        raise ValueError("review.verdict_type")
    if all(value[k] for k in FIELDS) != (value["reason"] == "none"):
        raise ValueError("review.reason_mismatch")
    return {k: value[k] for k in (*FIELDS, "reason")}


class ReviewSession:
    def __init__(self, *, case_timeout: float = 600, total_timeout: float = 1800):
        if not 0 < case_timeout <= 600 or not 0 < total_timeout <= 1800:
            raise ValueError("review.timeout_invalid")
        self.case_timeout = case_timeout
        self.remaining = total_timeout
        self.token = secrets.token_urlsafe(32)
        self.lock = threading.Condition()
        self.ready = False
        self.closed = False
        self.pending: dict[str, Any] | None = None
        self.decision: dict[str, Any] | None = None
        self.seen: set[str] = set()

    def state(self) -> dict[str, Any]:
        with self.lock:
            if self.closed:
                return dict(status="closed")
            if self.pending is not None:
                return dict(status="review", **self.pending)
            return dict(status="waiting" if self.ready else "ready_required")

    def mark_ready(self, value: Any) -> None:
        with self.lock:
            if value != {"ready": True} or type(value.get("ready")) is not bool:
                raise ValueError("review.ready_invalid")
            if self.closed or self.ready:
                raise ValueError("review.ready_repeated")
            self.ready = True
            self.lock.notify_all()

    def wait_ready(self) -> bool:
        with self.lock:
            self.lock.wait_for(lambda: self.ready or self.closed, timeout=self.case_timeout)
            return self.ready and not self.closed

    def submit(self, value: Any) -> None:
        with self.lock:
            if self.closed or self.pending is None or self.decision is not None:
                raise ValueError("review.not_pending")
            self.decision = verdict(value, self.pending["packet"]["caseId"], self.pending["nonce"])
            # The response is no longer retrievable once its single verdict is accepted.
            self.pending = None
            self.lock.notify_all()

    def review(self, value: dict[str, Any]) -> dict[str, Any]:
        raw = canonical(value)
        if len(raw) > MAX_PACKET_BYTES:
            raise ValueError("review.packet_oversize")
        case_id = value["caseId"]
        with self.lock:
            if not self.ready or self.closed or self.pending is not None or case_id in self.seen:
                raise ValueError("review.lifecycle_invalid")
            self.seen.add(case_id)
            self.decision = None
            self.pending = dict(nonce=secrets.token_urlsafe(24), packet=exact_json(raw))
            start = time.monotonic()
            try:
                self.lock.wait_for(lambda: self.decision is not None or self.closed,
                                   timeout=min(self.case_timeout, self.remaining))
                # A HTTP worker may have changed this while Condition released its lock.
                accepted = cast(dict[str, Any] | None, self.decision)
                result = dict(method="user_interactive", packetSha256=hashlib.sha256(raw).hexdigest(),
                              questionSha256=sha(value["question"]))
                if accepted is None:
                    return dict(**result, status="not_assessed", reason="closed" if self.closed else "timeout")
                return dict(**result, status="assessed", **accepted)
            finally:
                self.remaining = max(0, self.remaining - (time.monotonic() - start))
                self.pending = self.decision = None

    def close(self) -> None:
        with self.lock:
            self.closed = True
            self.pending = self.decision = None
            self.token = ""
            self.lock.notify_all()


class ReviewPortal:
    """Owns one temporary HTTP listener; static UI contains no credentials/data."""
    def __init__(self, session: ReviewSession):
        self.session = session
        self.server = ThreadingHTTPServer(("127.0.0.1", 0), self._handler())
        self.server.daemon_threads = True
        self.origin = f"http://127.0.0.1:{self.server.server_port}"
        self.thread = threading.Thread(target=self.server.serve_forever, daemon=True)

    def _handler(self) -> type[BaseHTTPRequestHandler]:
        owner = self

        class Handler(BaseHTTPRequestHandler):
            def setup(self) -> None:
                super().setup()
                self.connection.settimeout(3)

            def log_message(self, format: str, *args: Any) -> None:
                pass

            def send_body(self, code: int, body: bytes = b"", content_type: str = "application/json") -> None:
                self.send_response(code)
                self.send_header("Content-Type", content_type + "; charset=utf-8")
                self.send_header("Content-Length", str(len(body)))
                self.send_header("Cache-Control", "no-store")
                self.send_header("Referrer-Policy", "no-referrer")
                self.send_header("X-Content-Type-Options", "nosniff")
                self.send_header("Content-Security-Policy", "default-src 'none'; script-src 'self'; style-src 'self'; "
                    "connect-src 'self'; base-uri 'none'; frame-ancestors 'none'; form-action 'none'")
                self.end_headers()
                self.wfile.write(body)

            def allowed(self, *, private: bool) -> bool:
                if self.headers.get_all("Host") != [owner.origin.removeprefix("http://")]:
                    return False
                origins = self.headers.get_all("Origin")
                if origins is not None and origins != [owner.origin]:
                    return False
                if self.command == "POST" and origins != [owner.origin]:
                    return False
                if not private:
                    return True
                values = self.headers.get_all("X-Review-Token") or []
                token = owner.session.token
                return bool(token and len(values) == 1 and values[0].isascii() and hmac.compare_digest(values[0], token))

            def do_GET(self) -> None:
                if not self.allowed(private=self.path not in ("/", "/app.js", "/style.css")):
                    self.send_body(403)
                    return
                if self.path == "/state":
                    self.send_body(200, canonical(owner.session.state()))
                elif self.path in ("/", "/app.js", "/style.css"):
                    ext = {"/": "html", "/app.js": "js", "/style.css": "css"}[self.path]
                    content_type = {"html": "text/html", "js": "text/javascript", "css": "text/css"}[ext]
                    self.send_body(200, Path(__file__).with_suffix("." + ext).read_bytes(), content_type)
                else:
                    self.send_body(404)

            def do_POST(self) -> None:
                if not self.allowed(private=True):
                    self.send_body(403)
                    return
                if self.path not in ("/ready", "/decision", "/close"):
                    self.send_body(404)
                    return
                try:
                    lengths = self.headers.get_all("Content-Length") or []
                    if (len(lengths) != 1 or not lengths[0].isdigit()
                            or not 0 < int(lengths[0]) <= MAX_POST_BYTES
                            or self.headers.get_all("Content-Type") != ["application/json"]
                            or self.headers.get("Transfer-Encoding") is not None):
                        raise ValueError("review.body_invalid")
                    value = exact_json(self.rfile.read(int(lengths[0])))
                    if not isinstance(value, dict):
                        raise ValueError("review.body_shape")
                    if self.path == "/ready":
                        owner.session.mark_ready(value)
                    elif self.path == "/decision":
                        owner.session.submit(value)
                    else:
                        if value != {"close": True} or value["close"] is not True:
                            raise ValueError("review.close_invalid")
                        owner.session.close()
                    self.send_body(200, b'{"accepted":true}')
                except (ValueError, TypeError, KeyError, UnicodeError):
                    self.send_body(400)

        return Handler

    def __enter__(self) -> ReviewPortal:
        self.thread.start()
        return self

    def open_browser(self) -> bool:
        # No console URL or durable token file; fragment never reaches the HTTP server.
        return webbrowser.open(self.origin + "/#" + self.session.token, new=1)

    def __exit__(self, *args: Any) -> None:
        self.session.close()
        self.server.shutdown()
        self.server.server_close()
        self.thread.join(timeout=5)
